content-src/components/TopSites/SearchShortcutsForm.jsx

```diff
@@ -0,0 +1,143 @@
+        <label htmlFor={shortcut.keyword}>
+          <div className="top-site-inner">
+            <span>
+              <div className="tile">
+                <div className="top-site-icon rich-icon" style={imageStyle} data-fallback="@" />
+                <div className="top-site-icon search-topsite" />
+              </div>
+              <div className="title">
+                <span dir="auto">{shortcut.keyword}</span>
+              </div>
+            </span>
+          </div>
+        </label>
+      </div>
+    );
+  }
+}
+
+export class SearchShortcutsForm extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.handleChange = this.handleChange.bind(this);
+    this.onCancelButtonClick = this.onCancelButtonClick.bind(this);
+    this.onAddButtonClick = this.onAddButtonClick.bind(this);
+
+    // clone the shortcuts and add them to the state so we can add isSelected property
+    const shortcuts = [];
+    const {rows, searchShortcuts} = props.TopSites;
+    searchShortcuts.forEach(shortcut => {
+      shortcuts.push(Object.assign({}, shortcut, {isSelected: !!rows.find(({isPinned, searchTopSite, label}) => isPinned && searchTopSite && label === shortcut.keyword)}));
+    });
+    this.state = {shortcuts};
+  }
+
+  handleChange(event) {
+    const {target} = event;
+    const {name, checked} = target;
+    this.setState(prevState => {
+      const shortcuts = prevState.shortcuts.slice();
+      let shortcut = shortcuts.find(({keyword}) => keyword === name);
+      shortcut.isSelected = checked;
+      return {shortcuts};
+    });
+  }
+
+  onCancelButtonClick(ev) {
+    ev.preventDefault();
+    this.props.onClose();
+  }
+
+  onAddButtonClick(ev) {
+    ev.preventDefault();
+
+    // Check if there were any changes and act accordingly
+    const {rows} = this.props.TopSites;
+    const pinQueue = [];
+    const unpinQueue = [];
+    this.state.shortcuts.forEach(shortcut => {
+      const alreadyPinned = rows.find(({isPinned, searchTopSite, label}) => isPinned && searchTopSite && label === shortcut.keyword);
+      if (shortcut.isSelected && !alreadyPinned) {
+        pinQueue.push(this._searchTopSite(shortcut));
+      } else if (!shortcut.isSelected && alreadyPinned) {
+        unpinQueue.push({url: alreadyPinned.url});
+      }
+    });
+
+    // Pin the pinQueue
+    if (pinQueue.length > 0) {
+      // First find the available slots. A slot is available if it isn't pinned
+      // or if it's a pinned shortcut that we are about to unpin.
+      const availableSlots = [];
+      rows.forEach(({isPinned, searchTopSite, url}, index) => {
+        if (!isPinned || (searchTopSite && unpinQueue.find(site => url === site.url))) {
+          availableSlots.push(index);
+        }
+      });
+
+      pinQueue.forEach(shortcut => {
+        if (availableSlots.length === 0) {
+          // if they’ve pinned all their top sites then we do nothing
```

Comment:
This seems like it might result in unexpected behaviour for users; I wonder if there's a way we could  prevent this state?

Reply:
From our discussion with aaron, it sounds like the intended behaviour is to insert it at the first position and shift everything to the right